### PR TITLE
Fix mass matrix ForwardSensitivity

### PR DESCRIPTION
**Pull request by ArnoStrouwen**
Quick fix of mass matrix for ForwardSensitivity, but could probably be done more efficiently with Blockdiagonal matrices.

### CODE REVIEW

File: src/forward_sensitivity.jl [hunk 1]
```diff
@@ -175,9 +175,11 @@ function ODEForwardSensitivityProblem(f::F,u0,
     mm = f.mass_matrix
   else
     nn = size(f.mass_matrix, 1)
-    mm = similar(f.mass_matrix, 2nn, 2nn)
+    mm = zeros(eltype(f.mass_matrix), (length(p)+1)*nn, (length(p)+1)*nn)
     mm[1:nn, 1:nn] = f.mass_matrix
-    mm[nn+1:2nn, nn+1:2nn] = f.mass_matrix
+    for i = 1:length(p)
```
Comment (ChrisRackauckas) [inline src/forward_sensitivity.jl:180]:
```suggestion
    for i = 1:length(p)+1
```

Comment (ArnoStrouwen) [reply src/forward_sensitivity.jl:180]: That gives a dimension error.

The way I did it seems to pass tests on rober comparing:

ForwardDiffSensitivity with mass matrix
ForwardSensitivity with mass matrix
ForwardSensitivity with no mass matrix

Comment (ChrisRackauckas) [reply src/forward_sensitivity.jl:180]: oh I see, you're doing i+1

**Review by ChrisRackauckas — APPROVED** [review]

File: test/forward.jl [hunk 2]
```diff
@@ -152,7 +152,48 @@ tmp = similar(sol[1])
...
+
+prob_no_MM = ODEForwardSensitivityProblem(f_no_MM, u0, tspan, p, ForwardSensitivity())
+sol_no_MM= solve(prob_no_MM, Rodas5(autodiff = false), reltol = 1e-14, abstol = 1e-14)
+
+sen_MM_ForwardSensitivity = extract_local_sensitivities(sol_MM_ForwardSensitivity,10.0,true)
+sen_MM_ForwardDiffSensitivity = extract_local_sensitivities(sol_MM_ForwardDiffSensitivity,10.0,true)
+sen_no_MM = extract_local_sensitivities(sol_no_MM,10.0,true)
 
+@test sen_MM_ForwardSensitivity[2] ≈ sen_MM_ForwardDiffSensitivity[2] atol=1e-6 rtol=1e-6
+@test sen_MM_ForwardSensitivity[2] ≈ sen_no_MM[2] atol=1e-6 rtol=1e-6
```
Comment (ChrisRackauckas) [inline test/forward.jl:196]: why are these tolerances so high?

Comment (ArnoStrouwen) [reply test/forward.jl:196]: The tests fail when you make both atol and rtol moderately stringent. But if you make one less stringent then you can make the other one much more stringent.

I don't know what is best to do in this case?
In the test/forwaddiff.jl file plenty of the test don't both make atol and rtol stringent, only opting for one of the two.
```
@test sen_MM_ForwardSensitivity[2] ≈ sen_MM_ForwardDiffSensitivity[2] atol=1e-6 rtol=1e-6 # passes
@test sen_MM_ForwardSensitivity[2] ≈ sen_MM_ForwardDiffSensitivity[2] atol=1e-7 rtol=1e-7 # fails
@test sen_MM_ForwardSensitivity[2] ≈ sen_MM_ForwardDiffSensitivity[2] atol=1e-12 rtol=1e-6 # passes
@test sen_MM_ForwardSensitivity[2] ≈ sen_MM_ForwardDiffSensitivity[2] atol=1e-6 rtol=1e-12 # passes

@test sen_MM_ForwardSensitivity[2] ≈ sen_no_MM[2] atol=1e-6 rtol=1e-6 # passes
@test sen_MM_ForwardSensitivity[2] ≈ sen_no_MM[2] atol=1e-12 rtol=1e-6 # passes
@test sen_MM_ForwardSensitivity[2] ≈ sen_no_MM[2] atol=1e-6 rtol=1e-7 # fails
@test sen_MM_ForwardSensitivity[2] ≈ sen_no_MM[2] atol=1e-5 rtol=1e-12 # passes
```

Comment (ChrisRackauckas) [reply test/forward.jl:196]: interesting. Why is that the case?

Comment (ChrisRackauckas) [reply test/forward.jl:196]: Usually I just set one tolerance or thhe.

Comment (ArnoStrouwen) [reply test/forward.jl:196]: When passing both it has to pass either. Thus the test can not pass both atol and rtol more stringent than 1e-6.
The reason for this is likely the finite difference from `Rodas5(autodiff = false)` .
Getting the AD from the stiff solvers to play nice with sensitivity analysis is another can of worms.

Comment (ChrisRackauckas) [reply test/forward.jl:196]: Ahh, try something not Rosenbrock so it's less sensitive to AD.

**Comment by ChrisRackauckas** [comment]
Test failure seems real?

**Comment by ChrisRackauckas** [comment]
Rebase.

**Comment by ArnoStrouwen** [comment]
> Test failure seems real?

I doubt the issue is the changes to sensitivity code, since it is still fairly close.

```jl
Evaluated:
[-2.803577959425886 -1.1025166086946455e-9 6.6162921289797815e-6; 0.00010707093227508527 -1.90088870648834e-13 -4.828908182407572e-10; 2.8034754396259736 1.102710118418973e-9 -6.615829764690345e-6]
≈
[-2.8035779592733383 -1.102516609302266e-9 6.616292128269246e-6; 0.00010707093235252119 -1.9008887083111802e-13 -4.828908182976468e-10; 2.803468258412157 1.102704721479036e-9 -6.615797376473434e-6] (atol=1.0e-6, rtol=1.0e-6)

  Evaluated:
[-2.803577959425886 -1.1025166086946455e-9 6.6162921289797815e-6; 0.00010707093227508527 -1.90088870648834e-13 -4.828908182407572e-10; 2.8034754396259736 1.102710118418973e-9 -6.615829764690345e-6]
≈
[-2.8035779594262373 -1.1025166086959254e-9 6.61629212897741e-6; 0.0001070709322702509 -1.9008887072208858e-13 -4.828908182496889e-10; 2.803470888494033 1.1027066975666214e-9 -6.615809238159161e-6] (atol=1.0e-6, rtol=1.0e-6)
```

I can't reach stringent global tolerances using mass matrices:

```jl
using OrdinaryDiffEq
#using DiffEqSensitivity
using Test

function rober_MM(du, u, p, t)
    y₁, y₂, y₃ = u
    k₁, k₂, k₃ = p
    du[1] = -k₁ * y₁ + k₃ * y₂ * y₃
    du[2] = k₁ * y₁ - k₂ * y₂^2 - k₃ * y₂ * y₃
    du[3] = y₁ + y₂ + y₃ - 1
    nothing
end
function rober_no_MM(du, u, p, t)
    y₁, y₂, y₃ = u
    k₁, k₂, k₃ = p
    du[1] = -k₁ * y₁ + k₃ * y₂ * y₃
    du[2] = k₁ * y₁ - k₂ * y₂^2 - k₃ * y₂ * y₃
    du[3] = k₂*y₂^2
    nothing
end

M = [1.0 0 0; 0 1.0 0; 0 0 0]
p = [0.04, 3e7, 1e4]
u0 = [1.0, 0.0, 0.0]
tspan = (0.0, 12.0)

f_MM= ODEFunction(rober_MM, mass_matrix = M)
f_no_MM= ODEFunction(rober_no_MM)

prob1= ODEProblem(f_MM, u0, tspan, p)
sol1 = solve(prob1, Rodas5(autodiff=true), reltol = 1e-14, abstol = 1e-14)
prob2= ODEProblem(f_no_MM, u0, tspan, p)
sol2 = solve(prob2, Rodas5(autodiff=true), reltol = 1e-14, abstol = 1e-14)

@test sol1(10.0) ≈ sol2(10.0) atol=1e-5 # passes
@test sol1(10.0) ≈ sol2(10.0) atol=1e-6 # fails
@test sol1(10.0) ≈ sol2(10.0) rtol=1e-2 # passes
@test sol1(10.0) ≈ sol2(10.0) rtol=1e-3 # fails
```

**Comment by ChrisRackauckas** [comment]
Well of course, because using Rodas5 with mass matrices was a bad idea until... today. I'll put a tag in with https://github.com/SciML/OrdinaryDiffEq.jl/pull/1558 and https://github.com/SciML/OrdinaryDiffEq.jl/pull/1555 that would fix this, though before you would've needed to use something with a stiffly-aware interpolation like Rodas4

**Comment by ChrisRackauckas** [comment]
I put the OrdinaryDiffEq.jl tag in. Bump the lower bound if you want to keep Rodas5 in the tests there.

**Pull request merged**